group-coordinator/src/main/java/org/apache/kafka/coordinator/group/runtime/CoordinatorRuntime.java

```diff
@@ -370,6 +371,155 @@ public int size() {
+                snapshotRegistry.getOrCreateSnapshot(lastWrittenOffset);
+                snapshotRegistry.deleteSnapshotsUpTo(lastWrittenOffset);
+            }
+
+            /**
+             * Reverts the last written offset. This also reverts the snapshot
+             * registry to this offset. All the changes applied after the offset
+             * are lost.
+             *
+             * @param offset The offset to revert to.
+             */
+            private synchronized void revertLastWrittenOffset(
+                long offset
+            ) {
+                if (offset > lastWrittenOffset) {
+                    throw new IllegalStateException("New offset " + offset + " of " + tp +
+                        " must be smaller than " + lastWrittenOffset + ".");
+                }
+
+                log.debug("Revert last written offset of {} to {}.", tp, offset);
+                lastWrittenOffset = offset;
+                snapshotRegistry.revertToSnapshot(offset);
+            }
+
+            @Override
+            public void replay(
+                long producerId,
+                short producerEpoch,
+                U record
+            ) {
+                coordinator.replay(producerId, producerEpoch, record);
+            }
+
+            /**
+             * Updates the last written offset. This also create a new snapshot
+             * in the snapshot registry.
+             *
+             * @param offset The new last written offset.
+             */
+            @Override
+            public synchronized void updateLastWrittenOffset(Long offset) {
+                if (offset <= lastWrittenOffset) {
+                    throw new IllegalStateException("New last written offset " + offset + " of " + tp +
+                        " must be greater than " + lastWrittenOffset + ".");
+                }
+
+                log.debug("Update last written offset of {} to {}.", tp, offset);
+                lastWrittenOffset = offset;
+                snapshotRegistry.getOrCreateSnapshot(offset);
+            }
+
+            /**
+             * Updates the last committed offset. This completes all the deferred
+             * events waiting on this offset. This also cleanups all the snapshots
+             * prior to this offset.
+             *
+             * @param offset The new last committed offset.
+             */
+            @Override
+            public void updateLastCommittedOffset(Long offset) {
+                synchronized (this) {
+                    if (offset < lastCommittedOffset) {
+                        throw new IllegalStateException("New committed offset " + offset + " of " + tp +
+                            " must be greater than or equal to " + lastCommittedOffset + ".");
+                    }
+
+                    lastCommittedOffset = offset;
+                    deferredEventQueue.completeUpTo(offset);
+                    snapshotRegistry.deleteSnapshotsUpTo(offset);
```

Comment:
Accessing `deferredEventQueue` here is weird, no? It is owned by the CoordinatorContext and they it should be accessed while holding the lock of the context. We don't respect this here. I suppose that the same would apply to `coordinatorMetrics`.